ratis-logservice/src/main/java/org/apache/ratis/logservice/api/LogStream.java

```diff
@@ -0,0 +1,109 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.ratis.logservice.api;
+
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * A distributed log with "infinite" length that supports reads and writes.
+ */
+public interface LogStream {
+
+  /**
+   * Returns the unique name to identify this log.
+   */
+  LogName getName();
+
+  /**
+   * Returns the size of this LogStream in bytes.
+   */
+  long getSizeInBytes();
+
+  /**
+   * Returns the number of records in this LogStream.
+   */
+  long getSizeInRecords();
+
+  /**
+   * Creates a reader to read this LogStream asynchronously.
+   *
+   * @return An asynchronous reader
+   */
+  AsyncLogReader createAsyncReader();
+
+  /**
+   * Creates a writer to write to this LogStream asynchronously.
+   *
+   * @return An asynchronous writer
+   */
+  AsyncLogWriter createAsyncWriter();
+
+  /**
+   * Creates a reader to read this LogStream synchronously.
+   *
+   * @return A synchronous reader
+   */
+  LogReader createReader();
+
+  /**
+   * Creates a write to write to this LogStream synchronously.
+   *
+   * @return A synchronous writer
+   */
+  LogWriter createWriter();
+
+  /**
+   * Removes the elements in this LogStream prior to the given recordId.
+   *
+   * @param recordId A non-negative recordId for this LogStream
+   */
+  CompletableFuture<Void> truncateBefore(long recordId);
```

Comment:
Here also we should have a way whether to include the passed record id or not. If we don't want to include the passed recordid then we client of log service need to increment it.

Reply:
Do we really need this one? Use case?

Reply:
A LogStream is a sequence of entries: <recordId, byte[]> and RecordId is monotonically increasing. Truncation is the removal of entries from the "head" of the stream.

e.g. if we have a log with recordIds from `[0,100]`, we might say `truncateBefore(10, false)` which would remove entries 0, 1, 2, ..., 9. The "head" of the log would then be `recordId=10`